service/core/kafka/server/src/main/scala/com/lightbend/lagom/internal/broker/kafka/TopicProducerActor.scala

```diff
@@ -137,24 +133,23 @@ private[lagom] class TopicProducerActor[Message](
                       }
                     }
 
-                val eventPublisherFlow: Flow[(Message, AkkaOffset), Future[AkkaOffset], Any] =
-                  eventsPublisherFlow(endpoints, offset)
+                val eventPublisherFlow = eventsPublisherFlow(endpoints, offset)
 
                 // Return a Source[Future[Offset],_] where each produced element is a completed Offset.
                 eventStreamSource // read from DB + userFlow
-                  .map {
-                    case (message, offset) =>
-                      (
-                        message,
-                        ProjectionSpi.afterUserFlow(workerCoordinates.projectionName, workerCoordinates.tagName, offset)
-                      )
-                  }
+                  .map(command =>
+                    command.withOffset(
+                      ProjectionSpi
+                        .afterUserFlow(workerCoordinates.projectionName, workerCoordinates.tagName, command.offset)
+                    )
+                  )
                   .via(eventPublisherFlow) //  Kafka write + offset commit
-                  .map(_.map(offset => {
-                    ProjectionSpi
-                      .completedProcessing(workerCoordinates.projectionName, workerCoordinates.tagName, offset)
-                    offset
-                  }))
+                  .map(
+                    _.map(offset =>
+                      ProjectionSpi
+                        .completedProcessing(workerCoordinates.projectionName, workerCoordinates.tagName, offset)
+                    )
```

Comment:
Good catch! The instance returned by `completedProcessing` is not guaranteed to be the same as the one passed in. Your code is correct.